docs/icicle/primitives/poseidon.md

```diff
@@ -0,0 +1,228 @@
+# p = 0x12ab655e9a2ca55660b44d1e5c37b00159aa76fed00000010a11800000000001 # bls12-377
+# p = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001 # bn254
+# p = 0x1ae3a4617c510eac63b05c06ca1493b1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001 # bw6-761
+prime_bit_len = 255
+field_bytes = 32
+
+...
+
+# primitive_element = None
+primitive_element = 7 # bls12-381
+# primitive_element = 22 # bls12-377
+# primitive_element = 5 # bn254
+# primitive_element = 15 # bw6-761
+```
+
+We only support `alpha = 5` so if you want to use another alpha for SBox please reach out on discord or open a github issue.
+
+### Rust API
+
+This is the most basic way to use the Poseidon API.
+
+```rust
+let test_size = 1 << 10;
+let arity = 2u32;
+let ctx = get_default_device_context();
+let constants = load_optimized_poseidon_constants::<F>(arity, &ctx).unwrap();
+let config = PoseidonConfig::default();
+
+let inputs = vec![F::one(); test_size * arity as usize];
+let outputs = vec![F::zero(); test_size];
+let mut input_slice = HostOrDeviceSlice::on_host(inputs);
+let mut output_slice = HostOrDeviceSlice::on_host(outputs);
+
+poseidon_hash_many::<F>(
+    &mut input_slice,
+    &mut output_slice,
+    test_size as u32,
+    arity as u32,
+    &constants,
+    &config,
+)
+.unwrap();
+```
+
+The `PoseidonConfig::default()` can be modified, by default the inputs and outputs are set to be on `Host` for example.
+
+
+```
+impl<'a> Default for PoseidonConfig<'a> {
+    fn default() -> Self {
+        let ctx = get_default_device_context();
+        Self {
+            ctx,
+            are_inputs_on_device: false,
+            are_outputs_on_device: false,
+            input_is_a_state: false,
+            aligned: false,
+            loop_state: false,
+            is_async: false,
+        }
+    }
+}
+```
+
+In the example above `load_optimized_poseidon_constants::<F>(arity, &ctx).unwrap();` is used which will load the correct constants based on arity and curve. Its possible to [generate](#constants) your own constants and load them.
+
+```rust
+let ctx = get_default_device_context();
+    let cargo_manifest_dir = env!("CARGO_MANIFEST_DIR");
+    let constants_file = PathBuf::from(cargo_manifest_dir)
+        .join("tests")
+        .join(format!("{}_constants.bin", field_prefix));
+    let mut constants_buf = vec![];
+    File::open(constants_file)
+        .unwrap()
+        .read_to_end(&mut constants_buf)
+        .unwrap();
+
+    let mut custom_constants = vec![];
+    for chunk in constants_buf.chunks(field_bytes) {
+        custom_constants.push(F::from_bytes_le(chunk));
+    }
+
+    let custom_constants = create_optimized_poseidon_constants::<F>(
+        arity as u32,
+        &ctx,
+        full_rounds_half,
+        partial_rounds,
+        &mut custom_constants,
+    )
+    .unwrap();
+```
+For more examples using different configurations refer here.
+
+### Benchmarks 
+
+TODO
+
+
+## The Tree Builder
```

Comment:
Shouldn't it be on a different page?

Reply:
For now I think this is good, as we expand the tree builder more we will end up giving it a dedicated page.